scripts/fuzz_shell.js

```diff
@@ -423,16 +408,74 @@ function hashCombine(seed, value) {
       continue;
     }
 
+    // A task is a name + a function to call. For an export, the function is
+    // simply a call of the export.
+    tasks.push({ name: name, func: /* async */ () => callFunc(value) });
+  }
+
+  // Reverse the array, so the first task is at the end, for efficient
+  // popping in the common case.
+  tasks.reverse();
+
+  // Execute tasks while they remain.
+  while (tasks.length) {
+    let task;
+    if (ordering === undefined) {
+      // Use the natural order.
+      task = tasks.pop();
+    } else {
+      // Pick a random task.
+      ordering = hashCombine(ordering, tasks.length);
+      let i = ordering % tasks.length;
+      task = tasks.splice(i, 1)[0];
+    }
+
+    // Execute the task.
+    console.log('[fuzz-exec] calling ' + task.name);
+    let result;
     try {
-      console.log('[fuzz-exec] calling ' + name);
-      // TODO: Based on |ordering|, do not always await, leaving a promise
-      //       for later, so we interleave stacks.
-      var result = /* await */ callFunc(value);
-      if (typeof result !== 'undefined') {
-        console.log('[fuzz-exec] note result: ' + name + ' => ' + printed(result));
-      }
+      result = task.func();
     } catch (e) {
       console.log('exception thrown: ' + e);
+      continue;
+    }
+
+    if (JSPI) {
+      // When we are changing up the order, in JSPI we can also leave some
+      // promises unresolved until later, which lets us interleave them. Note we
+      // never defer a task more than once (which would be pointless), and we
+      // also only defer a promise (which we check for using .then).
+      if (ordering !== undefined && !task.deferred && result &&
+          typeof result == 'object' && typeof result.then === 'function') {
+        // Hash with -1 here, just to get something different than the hashing a
+        // few lines above.
+        ordering = hashCombine(ordering, -1);
+        if (ordering & 1) {
+          // Defer it for later. Reuse the existing task for simplicity.
+          console.log(`(jspi: defer ${task.name})`);
+          task.func = /* async */ () => {
```

Comment:
Since this is all in `if (JSPI)`, do the `async` and `await` need to be comments?

Reply:
My (perhaps overly paranoid) approach was then when JSPI is off, we don't want the JS VM to see any async stuff, which could affect codegen. Like maybe they lay out the stack or basic blocks differently in async functions?

Reply:
I would hope these isolated awaits/asyncs would not affect other code. Anway, I'm fine with leaving them since it's consistent with all the other uses.